network/pubsubchannel/pubsub.go

```diff
@@ -0,0 +1,159 @@
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+package pubsubchannel
+
+import (
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/swarm/log"
+)
+
+var subscriptionTimeout = 100 * time.Millisecond
+
+//PubSubChannel represents a pubsub system where subscriber can .Subscribe() and publishers can .Publish() or .Close().
+type PubSubChannel struct {
+	subscriptions []*Subscription
+	subsMutex     sync.RWMutex
+	nextId        int
+}
+
+// Subscription is created in PubSubChannel using pubSub.Subscribe(). Subscribers can receive using .ReceiveChannel().
+// or .Unsubscribe()
+type Subscription struct {
+	closed  bool
+	pubSubC *PubSubChannel
+	//removeSub func()
+	signal    chan interface{}
+	closeOnce sync.Once
+	id        string
+	lock      sync.RWMutex
+}
+
+// New creates a new PubSubChannel.
+func New() *PubSubChannel {
+	return &PubSubChannel{
+		subscriptions: make([]*Subscription, 0),
+	}
+}
+
+// Subscribe creates a subscription to a channel, each subscriber should keep its own Subscription instance.
+func (psc *PubSubChannel) Subscribe() *Subscription {
+	psc.subsMutex.Lock()
+	defer psc.subsMutex.Unlock()
+	newSubscription := newSubscription(strconv.Itoa(psc.nextId), psc)
+	psc.nextId++
+	psc.subscriptions = append(psc.subscriptions, &newSubscription)
+
+	return &newSubscription
+}
+
+func (psc *PubSubChannel) removeSub(s *Subscription) {
+	psc.subsMutex.Lock()
+	defer psc.subsMutex.Unlock()
+
+	for i, subscription := range psc.subscriptions {
+		if subscription.signal == s.signal {
+			log.Debug("Unsubscribing", "id", subscription.id)
+			subscription.lock.Lock()
+			subscription.closed = true
+			subscription.lock.Unlock()
+			psc.subscriptions = append(psc.subscriptions[:i], psc.subscriptions[i+1:]...)
+		}
+	}
+}
+
+// Publish broadcasts a message asynchronously to each subscriber.
+// If some of the subscriptions(channels) has been marked as closeable, it does it now.
+func (psc *PubSubChannel) Publish(msg interface{}) {
+	psc.subsMutex.RLock()
+	defer psc.subsMutex.RUnlock()
+	for _, sub := range psc.subscriptions {
+		sub.lock.RLock()
+		if sub.closed {
+			log.Debug("Subscription was closed", "id", sub.id)
+			sub.closeChannel()
+		} else {
+			go func(sub *Subscription) {
+				select {
+				case sub.signal <- msg:
+				case <-time.After(subscriptionTimeout):
```

Comment:
Is it ok to drop the message if subscriber gets slower then the timeout (100ms currently)? This is an important behaviour and it should be at least described in the comment on the Publish method so that clients know what functionality is provided. I think that it would be better to have a Subscription close channel (closed on unsubscribe) that would be responsible for returning from all goroutines created here for that subscription, and not to drop messages.

Reply:
Ok, let me rework this part.